posix/posix.c

```diff
@@ -2686,45 +2686,68 @@ int posix_waitpid(pid_t child, int *status, int options)
 
 void posix_died(pid_t pid, int exit)
 {
-	process_info_t *pinfo, *ppinfo, *init, *cinfo;
+	process_info_t *pinfo, *ppinfo, *init, *cinfo, *zinfo, *zombies = NULL, *children = NULL;
+	int wait;
 
-	pinfo = pinfo_find(pid);
-	if (pinfo == NULL) {
-		return;
-	}
+	init = pinfo_find(1);
+	LIB_ASSERT_ALWAYS(init != NULL, "init not found");
 
-	posix_exit(pinfo, exit);
+	pinfo = pinfo_find(pid);
+	LIB_ASSERT_ALWAYS(pinfo != NULL, "pinfo not found, pid: %d", pid);
 
 	ppinfo = pinfo_find(pinfo->parent);
-	if (ppinfo != NULL) {
-		proc_lockSet(&ppinfo->lock);
-		LIST_REMOVE(&ppinfo->children, pinfo);
-		LIST_ADD(&ppinfo->zombies, pinfo);
-		proc_threadBroadcast(&ppinfo->wait);
-		proc_lockClear(&ppinfo->lock);
+	LIB_ASSERT_ALWAYS(pinfo != NULL, "parent not found, pid: %d, ppid: %d", pid, pinfo->parent);
 
-		init = pinfo_find(1);
-		if (init != NULL) {
-			proc_lockSet2(&pinfo->lock, &init->lock);
-			while (pinfo->children != NULL) {
-				cinfo = pinfo->children;
+	posix_exit(pinfo, exit);
+
+	proc_lockSet(&pinfo->lock);
+	/* Collect all children */
+	children = pinfo->children;
+	pinfo->children = NULL;
+	/* Collect all zombies */
+	zombies = pinfo->zombies;
+	pinfo->zombies = NULL;
+	proc_lockClear(&pinfo->lock);
 
-				/* Treat cinfo->parent as atomic, no need for lock */
-				cinfo->parent = 1;
+	/* Make a zombie, wakeup waitpid */
+	proc_lockSet(&ppinfo->lock);
+	LIST_REMOVE(&ppinfo->children, pinfo);
+	if (ppinfo != init) {
+		LIST_ADD(&ppinfo->zombies, pinfo);
+		wait = proc_threadBroadcast(&ppinfo->wait);
+	}
+	else {
+		LIST_ADD(&zombies, pinfo);
+		wait = 1;
+	}
+	proc_lockClear(&ppinfo->lock);
 
-				LIST_REMOVE(&pinfo->children, cinfo);
-				LIST_ADD(&init->children, cinfo);
-			}
-			proc_lockClear(&pinfo->lock);
-			proc_lockClear(&init->lock);
-			pinfo_put(init);
+	/* Replant children */
+	if (children != NULL) {
+		proc_lockSet(&init->lock);
+		while (children != NULL) {
+			cinfo = children;
+			LIST_REMOVE(&children, cinfo);
+			cinfo->parent = 1;
```

Comment:
we've lost the comment about `pid_t` atomicity - might be important as `posix_died` is reentrant (current child might die in our hands) - if pid_t assignment would not be atomic, we might get invalid/mssing `ppinfo` for them.

If we're not sure about pid_t atomicity than maybe we need lock here?

EDIT:
Finally I think there might be some race conditions in reaping/reattaching zombies inside this function:
Assume we have process hierarchy INIT -> PARENT -> CHILD  (no grandchilds)

Scenario 1:
 - PARENT dies, is in `posix_died` in previous line (cinfo points to old parent)
 - CHILD dies, is recognized as zombie, is attached to `PARENT->zombies` list -> memory leak (parent pinfo would be reaped without re-checking zombies list)

Scenario 2:
 - PARENT dies, is in `posix_died` in next line (cinfo points to INIT as a parent)
 - CHILD dies, is recognized as zombie, is reaped at the end of the function -> we have entry in `INIT->children` pointing to deallocated memory

I think we might mitigate the race conditions with additional clever pinfo locking (eg. `pinfo->lock` should probably be held for the full duration of this function to ensure no new entries would be added to `pinfo->zombies`) and/or moving actual reaping to the init process or even making `posix_died` non-reentrant (drastic).

Reply:
You're right, thanks for spotting this.

Now I **replant** children and zombies in one operation, so race condition is no longer possible. Sadly, it means that proc_lockSet2 is needed again, but it doesn't hurt anything, I just wanted to avoid it, if possible

Reply:
Ok, one race is still possible, we might be adopted mid-death, will think about it

Reply:
Well, that was difficult, but I think I managed